src/g3log/logmessage.hpp

```diff
@@ -87,7 +88,8 @@ namespace g3 {
       // Complete access to the raw data in case the helper functions above
       // are not enough.
       //
-      std::time_t _timestamp;
+      // TODO: unique_ptr should be considered instead of shared_ptr
+      std::shared_ptr<Timestamp> _timestamp;
```

Comment:
I checked your solution with g_fatal_to_g3logworker_function_ptr global pointer. By default it points to pushFatalMessageToLogger. This is just an action that does not return any objects.

TimestampProvider does return an object so it is a different beast compared to your g_fatal_to_g3logworker_function_ptr. Moreover, different TimestampProvider's generate timestamps of different types (default provider returns std::time_t, in my case it is boost::posix_time::ptime), which is the the whole point of TimestampProvider API. Because timestamp objects can be of different types there are just two ways of saving them in the log messages:
-> Using pointers (this is what I did and it may have runtime performance impact)
-> Using templates (this is what can be done and it will have compile time impact, which I think is less painful)

Resolving how timestamps are saved in the relevant log message will lead to resolving other items you described below.